File: 2 Regression/D Support Vector/my_svr.r
#L68 SVR in R

#copypasta from pre template

# Regression Template

# Importing the dataset
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[2:3]

# Splitting the dataset into the Training set and Test set / not done here
# # install.packages('caTools')
# library(caTools)
# set.seed(123)
# split = sample.split(dataset$Salary, SplitRatio = 2/3)
# training_set = subset(dataset, split == TRUE)
# test_set = subset(dataset, split == FALSE)

# Feature Scaling  -NOTE NOT NEEDED HERE? (NEEDED IN PY W SVR)
# training_set = scale(training_set)
# test_set = scale(test_set)

# Fitting the SVR Model to the dataset / !!! lecture 68 from here !!!
install.packages('e1071')
library(e1071)
regressor = svm(formula = Salary ~ ., 
                data = dataset,
                type = 'eps-regression')

# Predicting a new result
y_pred = predict(regressor, data.frame(Level = 6.5))

# Visualising the Regression Model results
# install.packages('ggplot2')
library(ggplot2)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red') +
  geom_line(aes(x = dataset$Level, y = predict(regressor, newdata = dataset)),
            colour = 'blue') +
  ggtitle('Truth or Bluff (SVR Model)') +
  xlab('Level') +
  ylab('Salary')

# Visualising the Regression Model results (for higher resolution and smoother curve)
# install.packages('ggplot2')
library(ggplot2)
x_grid = seq(min(dataset$Level), max(dataset$Level), 0.1)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red') +
  geom_line(aes(x = x_grid, y = predict(regressor, newdata = data.frame(Level = x_grid))),
            colour = 'blue') +
  ggtitle('Truth or Bluff (SVR Model)') +
  xlab('Level') +
  ylab('Salary')

File: 2 Regression/D Support Vector/my_svr.py
# -*- coding: utf-8 -*-
"""
Created on Wed Jun 20 07:52:40 2018

@author: USER
"""

# lecture 67 - SVR from template
# SVR disregards errors within epsilon value

# Regression Template

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Position_Salaries.csv')
X = dataset.iloc[:, 1:2].values
y = dataset.iloc[:, 2:3].values
# need to reshape Y array for feature scaling so changed col from 2 to 2:3
# possible alt is y = y.reshape(-1,1)


# Splitting the dataset into the Training set and Test set
"""from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)"""

# Feature Scaling -- NEEDED FOR SVR (aabout 6:30, L67)
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
sc_y = StandardScaler()
X = sc_X.fit_transform(X)
y = sc_y.fit_transform(y)
# gives "change int 64 to float 64" warning.  no biggie

# !!!
# Fitting thSVR to dataset
# !!!

from sklearn.svm import SVR
regressor = SVR(kernel = 'rbf')
# kernel: linear, polynomial, gaussian, rbf (default but good to specify)
# also set epsilon val, etc.
regressor.fit(X, y)
# gives output

# !!! that's it!

# Predicting a new result / edits made around 15 -1 7 mark
y_pred = sc_y.inverse_transform(regressor.predict(sc_X.transform(np.array([[6.5]])))) 
# need to change 6.5 to transformed value b/c of feature scaling
# but transform expects an array so 6.5 cannot be data type
#use numpy array to create 1 cell array, two pairs of brackets makes array 1x1 (one pair = vector)
# need to use inverse transform on y to then get orig scale of salary



# Visualising the Regression results
plt.scatter(X, y, color = 'red')
plt.plot(X, regressor.predict(X), color = 'blue')
plt.title('Truth or Bluff (Regression Model)')
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show()

# or can use "smoother" code for smoother curve
